File: day15/generators.c
#include <stdio.h>

int main() {
  long int a = 783;
  long int b = 325;
  
  int fa = 16807;
  int fb = 48271;
  int p = 2147483647;

  int count = 0;
  for (int i = 0; i < 40000000; ++i) {
    a = (a * fa) % p;
    b = (b * fb) % p;
    /* printf("%ld\t%ld\n", a, b); */
    if ((a & 0xFFFF) == (b & 0xFFFF)) {
      ++count;
    }
  }
  printf("%d\n", count);
}

File: day15/generators.hs
import Data.Bits as B ((.&.))
import Data.List (foldl')

a = 16807 :: Int
b = 48271 :: Int
p = 2147483647

generate :: Int -> Int -> [Int]
generate factor seed = map (.&. 0xFFFF) $ tail $ iterate (\x -> rem (x * factor) p) seed

countMatching n seedA seedB =
  let numbers = zipWith (==) (take n $ generate a seedA) (take n $ generate b seedB)
  in foldl' (\acc v -> if v then acc + 1 else acc) 0 numbers
                      
main = putStrLn $ show $ countMatching 40000000 783 325
